apps/builder/app/builder/features/settings-panel/controls/resource-control.tsx

```diff
@@ -0,0 +1,265 @@
+    const systemIdentifier = encodeDataSourceVariable(page.systemDataSourceId);
+    delete newScope[systemIdentifier];
+    newAliases.delete(systemIdentifier);
+    return { scope: newScope, aliases: newAliases };
+  }
+);
+
+const ResourceForm = ({ resource }: { resource: undefined | Resource }) => {
+  const bindingPopoverContainerRef = useRef<HTMLDivElement>(null);
+  // @todo exclude collection item and system
+  // basically all parameter variables
+  const { scope, aliases } = useStore($scope);
+  const [url, setUrl] = useState(resource?.url ?? `""`);
+  const [method, setMethod] = useState<Resource["method"]>(
+    resource?.method ?? "post"
+  );
+  const [headers, setHeaders] = useState<Resource["headers"]>(
+    resource?.headers ?? []
+  );
+  return (
+    <Flex
+      ref={bindingPopoverContainerRef}
+      direction="column"
+      css={{
+        width: theme.spacing[30],
+        overflow: "hidden",
+        gap: theme.spacing[9],
+        p: theme.spacing[9],
+      }}
+    >
+      <BindingPopoverProvider
+        value={{ containerRef: bindingPopoverContainerRef }}
+      >
+        <UrlField
+          scope={scope}
+          aliases={aliases}
+          value={url}
+          onChange={setUrl}
+          onCurlPaste={(curl) => {
+            // update all feilds when curl is paste into url field
+            setUrl(JSON.stringify(curl.url));
+            setMethod(curl.method);
+            setHeaders(
+              curl.headers.map((header) => ({
+                name: header.name,
+                value: JSON.stringify(header.value),
+              }))
+            );
+          }}
+        />
+        <MethodField value={method} onChange={setMethod} />
+        <Headers
+          scope={scope}
+          aliases={aliases}
+          headers={headers}
+          onChange={setHeaders}
+        />
+      </BindingPopoverProvider>
+    </Flex>
+  );
+};
+
+const setResource = ({
+  instanceId,
+  propId,
+  propName,
+  resource,
+}: {
+  instanceId: Prop["instanceId"];
+  propId?: Prop["id"];
+  propName: Prop["name"];
+  resource: Resource;
+}) => {
+  updateWebstudioData((data) => {
+    setPropMutable({
+      data,
+      update: {
+        id: propId ?? nanoid(),
+        instanceId,
+        name: propName,
+        type: "resource",
+        value: resource.id,
+      },
+    });
+    data.resources.set(resource.id, resource);
+  });
+};
+
+const areAllFormErrorsVisible = (form: null | HTMLFormElement) => {
+  if (form === null) {
+    return true;
+  }
+  // check all errors in form fields are visible
+  for (const element of form.elements) {
+    if (
+      element instanceof HTMLInputElement ||
+      element instanceof HTMLTextAreaElement
+    ) {
+      // field is invalid and the error is not visible
+      if (
+        element.validity.valid === false &&
+        // rely on data-color=error convention in webstudio design system
+        element.getAttribute("data-color") !== "error"
```

Comment:
looks like dirty hack

Reply:
It is though it works and I didn't figure out another way.